core/mutator/entry/mutator_test.go

```diff
@@ -30,82 +30,155 @@ func TestCheckMutation(t *testing.T) {
+		}, // Second mutation, working case
+		{
+			old: entryData2,
+			mutation: &Mutation{
+				index: key,
+				entry: &tpb.Entry{
+					Commitment:     []byte{2},
+					Previous:       hashEntry1[:],
+					AuthorizedKeys: mustPublicKeys([]string{testPubKey2}),
+				},
+			},
+			err: mutator.ErrReplay,
+		}, // Replayed mutation
+		{
+			old: entryData1,
+			mutation: &Mutation{
+				index: bytes.Repeat(key, mutator.MaxMutationSize),
+			},
+			err: mutator.ErrSize,
+		}, // Large mutation
+		{
+			old: entryData2,
+			mutation: &Mutation{
+				index: key,
+				entry: &tpb.Entry{
+					Commitment:     []byte{2},
+					Previous:       nil,
+					AuthorizedKeys: mustPublicKeys([]string{testPubKey2}),
+				},
+			},
+			err: mutator.ErrPreviousHash,
+		}, // Invalid previous entry hash
+		{
+			mutation: &Mutation{
+				entry: &tpb.Entry{
+					Previous: nil,
+				},
+			},
+			err: mutator.ErrPreviousHash,
+		}, // Very first mutation, invalid previous entry hash
+		{
+			mutation: &Mutation{
+				index: key,
+				entry: &tpb.Entry{
+					Commitment:     []byte{2},
+					Previous:       nilHash[:],
+					AuthorizedKeys: mustPublicKeys([]string{}),
+				},
+			},
+			err:     mutator.ErrMissingKey,
+			signers: signersFromPEMs(t, [][]byte{[]byte(testPrivKey1)}),
+		}, // Very first mutation, missing current key
+		{
+			mutation: &Mutation{
+				index: key,
+				entry: &tpb.Entry{
+					Commitment:     []byte{2},
+					Previous:       nilHash[:],
+					AuthorizedKeys: mustPublicKeys([]string{testPubKey1}),
+				},
+			},
+			signers: signersFromPEMs(t, [][]byte{}),
+			err:     mutator.ErrUnauthorized,
+		}, // Very first mutation, missing current signature
+		{
+			old: entryData1,
+			mutation: &Mutation{
+				index: key,
+				entry: &tpb.Entry{
+					Commitment:     []byte{2},
+					Previous:       hashEntry1[:],
+					AuthorizedKeys: mustPublicKeys([]string{testPubKey2}),
+				},
+			},
+			signers: signersFromPEMs(t, [][]byte{[]byte(testPrivKey2)}),
+			err:     mutator.ErrUnauthorized,
+		},
+		{
+			old: entryData1,
+			mutation: &Mutation{
+				index: key,
+				entry: &tpb.Entry{
+					Commitment:     []byte{2},
+					Previous:       hashEntry1[:],
+					AuthorizedKeys: mustPublicKeys([]string{testPubKey2}),
+				},
+			},
+			signers: signersFromPEMs(t, [][]byte{[]byte(testPrivKey1)}),
+		},
 	} {
-		// Prepare mutations.
-		m := &Mutation{
-			index: tc.key,
-			entry: tc.newEntry,
-		}
-		m.entry.Previous = tc.previous
-		mutation, err := m.sign(tc.signers)
+		m, err := tc.mutation.sign(tc.signers)
 		if err != nil {
 			t.Errorf("mutation.sign(%v): %v", tc.signers, err)
 			continue
 		}
 
-		if _, got := New().Mutate(tc.oldEntry, mutation); got != tc.err {
-			t.Errorf("%d Mutate(%v, %v): %v, want %v", i, tc.oldEntry, mutation, got, tc.err)
+		if _, got := New().Mutate(tc.old, m); got != tc.err {
+			t.Errorf("%d Mutate(): %v, want %v", i, got, tc.err)
```

Comment:
Not really your diff, but having a `desc string` field in the test table gives more easily understood output than a test case index when something goes wrong.

Reply:
done